File: packages/core/src/envelope.ts
// The Envelope: the thin universal waist. Every Aleph message is a signed,
// addressed, stateless Envelope between two DIDs, carrying one of five types.
// This is the "IP packet" of the agentic web. Signatures are domain-separated
// (an envelope signature cannot be reused as a grant/attestation/…).

import { randomUUID } from "node:crypto";
import { type Identity } from "./identity";
import { DOMAIN, signWith, verifyByDid, type Signer } from "./signing";

export const PROTOCOL_VERSION = "aleph/0.1";

export type EnvelopeType = "RESOLVE" | "INVOKE" | "RECEIPT" | "ATTEST" | "SETTLE";

export interface Envelope {
  v: string;
  from: string;
  to: string;
  type: EnvelopeType;
  nonce: string;
  ts: number;
  body: Record<string, unknown>;
  sig?: string;
}

export function createEnvelope(
  params: { from: string; to: string; type: EnvelopeType; body: Record<string, unknown> },
  // Ed25519 key (the common case) or any Signer (secp256k1 did:key / did:pkh).
  privateKey: Identity["privateKey"] | Signer,
): Envelope {
  // Build the unsigned Envelope, sign its domain-separated canonical form.
  const env: Envelope = {
    v: PROTOCOL_VERSION,
    from: params.from,
    to: params.to,
    type: params.type,
    nonce: randomUUID(),
    ts: Date.now(),
    body: params.body,
  };
  env.sig = signWith(privateKey, DOMAIN.envelope, env);
  return env;
}

export function verifyEnvelope(env: Envelope): { ok: boolean; reason?: string } {
  if (!env.sig) return { ok: false, reason: "missing signature" };
  if (env.v.split("/")[0] !== "aleph") return { ok: false, reason: "unknown protocol family" };
  const { sig, ...unsigned } = env;
  try {
    // Suite-agnostic: the signer's DID declares Ed25519 or secp256k1.
    const ok = verifyByDid(env.from, DOMAIN.envelope, unsigned, sig);
    return ok ? { ok: true } : { ok: false, reason: "bad signature" };
  } catch (e) {
    return { ok: false, reason: (e as Error).message };
  }
}

File: sdk/python/aleph_protocol/envelope.py
"""The Envelope — the universal signed message between two DIDs. Reproduces
@aleph/core's envelope.ts: the unsigned envelope is signed over the `envelope`
domain; `from` is the signer's DID."""

import time
import uuid

from .identity import Identity
from .signing import DOMAIN, sign_ed25519, verify_by_did

PROTOCOL_VERSION = "aleph/0.1"

# RESOLVE / INVOKE / RECEIPT / ATTEST / SETTLE
EnvelopeType = str


def create_envelope(
    sender: Identity,
    to: str,
    type: EnvelopeType,
    body: dict,
    *,
    nonce: str = None,
    ts: int = None,
) -> dict:
    env = {
        "v": PROTOCOL_VERSION,
        "from": sender.did,
        "to": to,
        "type": type,
        "nonce": nonce or str(uuid.uuid4()),
        "ts": ts if ts is not None else int(time.time() * 1000),
        "body": body,
    }
    env["sig"] = sign_ed25519(DOMAIN["envelope"], env, sender.private_key)
    return env


def verify_envelope(env: dict) -> bool:
    if "sig" not in env or not env["sig"]:
        return False
    unsigned = {k: v for k, v in env.items() if k != "sig"}
    return verify_by_did(env["from"], DOMAIN["envelope"], unsigned, env["sig"])
